src/Microsoft.AspNet.OData.Shared/Extensions/ODataPathExtensions.cs

```diff
@@ -11,46 +11,91 @@
 
 namespace Microsoft.AspNet.OData.Extensions
 {
+    /// <summary>
+    /// Extensions method for <see cref="ODataPath"/>.
+    /// </summary>
     internal static class ODataPathExtensions
     {
-        public static Dictionary<string, object> GetKeys(this ODataPath path)
+        /// <summary>
+        /// Get keys from the last <see cref="KeySegment"/>.
+        /// </summary>
+        /// <param name="path"><see cref="ODataPath"/>.</param>
+        /// <returns>Dictionary of keys.</returns>
+        internal static Dictionary<string, object> GetKeys(this ODataPath path)
         {
             if (path == null)
             {
                 throw Error.ArgumentNull(nameof(path));
             }
 
-            Dictionary<string, object> keys = new Dictionary<string, object>();
+            List<ODataPathSegment> pathSegments = path.AsList();
 
-            // Books(1)/Authors(1000)/Namespace.SpecialAuthor
-            if (path.LastSegment is TypeSegment)
+            if (pathSegments.Count == 0)
             {
-                ODataPath pathWithoutLastSegmentCastType = path.TrimEndingTypeSegment();
-
-                if (pathWithoutLastSegmentCastType.LastSegment is KeySegment)
-                {
-                    keys = GetKeysFromKeySegment(pathWithoutLastSegmentCastType.LastSegment as KeySegment);
-                }
+                return null;
```

Comment:
Is there any reason we changed the method to return `null` instead of empty dictionary like before when the `KeySegment` isn't present in `pathSegments`?

Reply:
Actual there is no reason.
Probably we don't want to allocate a dictionary object in memory whenever we don't have a key segment

Reply:
I think that's a fair argument. Another option could be to use a `TryGetKeys` with an `out` parameter for the dictionary and a `bool` return for whether or not there were keys. But I'm okay with all approaches.